### PR TITLE
render: Implement more PixelBender features

### PR DESCRIPTION
**Pull request by Aaron1011**
This PR contains several inter-related changes:

* Implement ShaderFilter, and connect it to PixelBender execution
* Implement PixelBender matrix parameters and local variables. These are passed in as floating point parameters through the existing float parameters uniform. Howevever, they are a distinct class of local variables (like int variables), and receive distinct naga local variable declarations.
* Implement several more PixelBender opcodes, including matrix multiplication.

When combined with some other unrelated PRs, this gets John Citizen 2 working (including lighting and ice reflection PixelBender shaders).

### CODE REVIEW

File: core/src/avm2/filters.rs [hunk 1]
```diff
@@ -105,6 +116,9 @@ impl FilterAvm2Ext for Filter {
                 let gradientglowfilter = activation.avm2().classes().gradientglowfilter;
                 gradient_filter_to_avm2(activation, filter, gradientglowfilter)
             }
+            Filter::ShaderFilter(_) => Err(Error::RustError(
+                "Shader filter conversion not yet implemented".into(),
```
Comment (Dinnerbone) [inline core/src/avm2/filters.rs:120]: Is it possible to implement this? I don't know how the impact would be, but anything that has a shader filter will now error whenever you access the filter list

File: core/src/avm2/globals/flash/display/shader_job.rs [hunk 2]
```diff
@@ -108,31 +99,63 @@ pub fn start<'gc>(
...
+
+                        // FIXME - this really only needs to be a CPU->GPU sync
+                        let bitmap = bitmap.sync();
+                        let mut bitmap_data = bitmap.write(activation.context.gc_context);
+                        bitmap_data.update_dirty_texture(activation.context.renderer);
+                        Some(
+                            bitmap_data
+                                .bitmap_handle(activation.context.renderer)
+                                .expect("Missing input BitmapHandle"),
+                        )
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display/shader_job.rs:122]:
```suggestion
                        let handle = bitmap.bitmap_handle(activation.context.gc_context, activation.context.renderer);
                        Some(
                            handle
                                .expect("Missing input BitmapHandle"),
                        )
```

File: render/wgpu/src/backend.rs [hunk 3]
```diff
@@ -917,7 +922,61 @@ impl<T: RenderTarget + 'static> RenderBackend for WgpuRenderBackend<T> {
...
+            width: target.texture.width(),
+            height: target.texture.height(),
+            depth_or_array_layers: 1,
+        };
+
+        let mut texture_target = TextureTarget {
+            size: extent,
+            texture: target.texture.clone(),
+            format: wgpu::TextureFormat::Rgba8Unorm,
+            buffer: None,
```
Comment (Dinnerbone) [inline render/wgpu/src/backend.rs:937]: It'd be nice to get the automatic buffer logic here too, for faster readbacks

File: tests/tests/swfs/avm2/pixelbender_effect_glassDisplace/test.toml [hunk 4]
```diff
@@ -2,7 +2,7 @@ num_frames = 1
 
 [image_comparison]
 tolerance = 1
-max_outliers = 34
+max_outliers = 1640
```
Comment (Dinnerbone) [inline tests/tests/swfs/avm2/pixelbender_effect_glassDisplace/test.toml:5]: This may not need to be this high, try just increasing the tolerance by 1 or 2 and use outliers for the really big outliers

File: tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/test.toml [hunk 5]
```diff
@@ -2,6 +2,7 @@ num_frames = 10
 
 [image_comparison]
 tolerance = 1
+max_outliers = 484
```
Comment (Dinnerbone) [inline tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/test.toml:5]: Same here

File: core/src/avm2/globals/flash/display/BitmapData.as [hunk 6]
```diff
@@ -68,6 +69,10 @@ package flash.display {
         ):int;
 
         public function generateFilterRect(sourceRect:Rectangle, filter:BitmapFilter):Rectangle {
+            // Flash always reports that a ShaderFilter affects the entire BitampData, ignoring SourceRect.
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/BitmapData.as:72]: Can you add a test for this in BitmapData? Also, does this ignore the ShaderFilter extensions?

File: tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/YellowFlowers.png
Comment (Lord-McSweeney) [inline tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/YellowFlowers.png:1]: Accidentally included? This doesn't seem relevant to the test.

File: render/naga-pixelbender/src/lib.rs [hunk 7]
```diff
@@ -330,25 +523,38 @@ impl<'a> ShaderBuilder<'a> {
...
                         | PixelBenderTypeOpcode::TInt3
                         | PixelBenderTypeOpcode::TInt4 => {
                             num_ints += 1;
-                            (int_offset, false)
+                            (int_offset, ParamKind::Int)
                         }
                         PixelBenderTypeOpcode::TString => continue,
-                        _ => unimplemented!("Unsupported parameter type {:?}", param_type),
+                        PixelBenderTypeOpcode::TFloat2x2 => {
+                            num_floats += 1;
```
Comment (Lord-McSweeney) [inline render/naga-pixelbender/src/lib.rs:545]: I'm not sure if this is correct, but should this be a 2 instead of 1?

File: core/src/avm2/globals/flash/display/BitmapData.as (hunk 6 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/display/BitmapData.as:72]: There's already a test for this in `pixelbender_effect_glassDisplace_shaderfilter`

File: tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/YellowFlowers.png
Comment (Aaron1011) [reply tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/YellowFlowers.png:1]: This is imported by the .fla

File: render/wgpu/src/backend.rs [hunk 8]
```diff
@@ -917,7 +919,66 @@ impl<T: RenderTarget + 'static> RenderBackend for WgpuRenderBackend<T> {
...
+            }),
+            // When running a standalone shader, we always process the entire image
+            &FilterSource::for_entire_texture(&target.texture),
+        )?;
+
+        self.descriptors
+            .queue
+            .submit(Some(render_command_encoder.finish()));
+
+        Ok(Box::new(QueueSyncHandle::NotCopied {
```
Comment (Dinnerbone) [inline render/wgpu/src/backend.rs:976]: This now needs to be `QueueSyncHandle::AlreadyCopied` depending on target buffer

File: tests/tests/swfs/avm2/pixelbender_effect_glassDisplace/test.toml (hunk 4 above)
Comment (Aaron1011) [reply tests/tests/swfs/avm2/pixelbender_effect_glassDisplace/test.toml]: Bumping tolerance to 3 helped a little, but there's still a large number of outliers, unfortunately.

File: tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/test.toml (hunk 5 above)
Comment (Aaron1011) [reply tests/tests/swfs/avm2/pixelbender_effect_glassDisplace_shaderfilter/test.toml]: Same result as in the other test, unfortunately

File: core/src/avm2/globals/flash/display/shader_job.rs [hunk 9]
```diff
@@ -108,31 +99,58 @@ pub fn start<'gc>(
...
+                        let input = input
+                            .as_object()
+                            .expect("ShaderInput.input is not an object");
 
-                    // FIXME - this really only needs to be a CPU->GPU sync
-                    let bitmap = bitmap.sync();
-                    let mut bitmap_data = bitmap.write(activation.context.gc_context);
-                    bitmap_data.update_dirty_texture(activation.context.renderer);
+                        let bitmap = input.as_bitmap_data().expect(
+                            "ShaderInput.input is not a BitmapData (FIXE - support other types)",
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display/shader_job.rs:111]: nit: FIXE --> FIXME

File: core/src/avm2/globals/flash/filters/ShaderFilter.as [hunk 10]
```diff
@@ -15,6 +15,11 @@ package flash.filters {
         public function set shader(value:Shader):void {
             this._shader = value;
         }
+
+        public var bottomExtension:int = 0;
+        public var leftExtension:int = 0;
+        public var rightExtension:int = 0;
+        public var topExtension:int = 0;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/filters/ShaderFilter.as:22]: AFAIK FP has constraints on these and uses a getter/setter for them. While figuring out the constraints might take some reverse-engineering (since we can't look at the playerglobal code), for consistency and later use these should probably be made into getters and setters.

File: core/src/avm2/globals/flash/filters/ShaderFilter.as [hunk 11]
```diff
@@ -16,10 +16,42 @@ package flash.filters {
...
+        public function set leftExtension(value:int):void {
+            this._leftExtension = value;
+        }
+
+        public function get rightExtension():int {
+            return this._rightExtension;
+        }
+
+        public function set rightExtension(value:int):void {
+            this._ = value;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/filters/ShaderFilter.as:45]: typo here: _ --> _rightExtension
Also, the tests didn't catch this- are there no tests for the extensions currently?

Comment (Aaron1011) [reply core/src/avm2/globals/flash/filters/ShaderFilter.as]: They're not hooked up to anything yet, so I haven't written any tests.

**Review by Dinnerbone — APPROVED** [review]
